analysis/subset_data/SubsetBam.wdl

```diff
@@ -0,0 +1,169 @@
+workflow SubsetBam {
+  # Bam to subset
+  File aligned_gold_bam
+
+  # How many cells to grab from the largest and smallest end
+  Int num_large_cells
+  Int num_small_cells
+
+  # How many reads a cell must have before it can be selected
+  Int? min_reads_per_cell
+
+  # Array of comma separated tag values to put in each fastq e.g ["ORIG", "SR", "CB,UB"] "ORIG" means to
+  # grab the original sequence/qualities from the read
+  Array[String] barcode_tags_to_split_by
+  Array[String] quality_tags_to_split_by
+  Array[String]? separators_to_split_by
+
+  Array[String]? final_fastq_basenames
+
+  Int bam_size = ceil(size(aligned_gold_bam, "GB"))
+  Int additional_disk = 10
+
+  # Figure out which cell barcodes to select from bam
+  call SelectCells {
+    input:
+      aligned_bam = aligned_gold_bam,
+      num_large_cells = num_large_cells,
+      num_small_cells = num_small_cells,
+      min_reads_per_cell = min_reads_per_cell,
+      disk = bam_size + additional_disk
```

Comment:
There are a few differences.  For this workflow the only real benefit is `additional_disk` (wiggle room disk) that could be passed to each task from one value instead of each task having another input for the same purposes.  For other workflows I've written where like alot of the tools use a reference fasta, fasta index, and dictionary you can package the size of that package into one variable and use that to help calculate the size for multiple workflows instead of having 3 size calculations in each of those tasks.  Another reason that isn't used in this workflow is for tasks use streaming directly from the cloud (NIO) within a scatter, only the workflow would have an idea of the size expected from each shard naturally (you could pass that value into the task but again just more repeated things).  I definitely agree that task level size calculations are more in line with the one wdl, one task philosophy (which as I said earlier I love) but I think it really depends on the situation.

One option is to have a size calculation in the task but also allow the task to take in an optional disk_size input so the user can use whichever way suits their needs?  What do you think

Reply:
Thanks for the explanation. 100% agree w.r.t. NIO, obviously you need to pre-allocate there. 

I really like the last idea; that said it is also the most work for the developer...